### PR TITLE
Add gru batch logic (forward and back prop)

### PR DESCRIPTION
**Pull request by catli**
Clean up the code around batching sessions for forward and back propagation.

### CODE REVIEW

File: model/gru.py [hunk 1]
```diff
@@ -46,21 +46,36 @@ def init_hidden(self):
...
         '''
             apply the forward function for the model
+            clarifying with example from:
+            towardsdatascience.com/\
+            taming-lstms-variable-sized-mini-batches-and-why-pytorch-is-good-for-your-health-61d35642972e
         '''
-        seq_len = batch_data.size()[1]
-        # check that the gru unit treats the s
-        gru_out, self.hidden = self.model(batch_data, self.hidden)
-        output = self.hidden_to_output(gru_out)
```
Comment (catli) [inline model/gru.py:59]: @zhampel whenever you have a chance it would be great to get your thoughts on how I structure this GRU model. 

I'm organizing the forward prop as follows:    

   gru_out =  GRU(input)
   linear_output = Linear(gru_out)
   output = sigmoid(linear_output)

File: model/gru.py [hunk 2]
```diff
@@ -46,21 +46,36 @@ def init_hidden(self):
...
+        # [TODO] check to see what this unpacking is doing
+        unpacked_out = unpacked_out.contiguous()
+        unpacked_out = unpacked_out.view(-1, unpacked_out.shape[2])
+        unpacked_out = self.hidden_to_output(unpacked_out)
+        # Add a sigmoid layer
+        sigmoid_out = torch.sigmoid(unpacked_out)
+        output = sigmoid_out.view(self.batch_size, -1, self.output_dim)
         return output
 
     def loss(self, output, label):
```
Comment (catli) [inline model/gru.py:77]: @zhampel I have a placeholder loss function here, but what kind of loss would you recommending using and would be most effective if each label is multi-hot vector
[ 0,1,...1,...0]

and the output has the likelihoods.

**Comment by catli** [comment]
New updates to effectively run batched data.

File: model/gru.py (hunk 2 above)
Comment (zhampel) [reply model/gru.py:77]: @catli (sorry it took a few days, finally back from travel) For this I think you want to use a multi-class cross entropy type loss function. It's an extension/generalization of binary cross entropy to multiple classes. You can look [here](https://ml-cheatsheet.readthedocs.io/en/latest/loss_functions.html) for a defn, and maybe [here](https://discuss.pytorch.org/t/multi-class-cross-entropy-loss-function-implementation-in-pytorch/19077) for an example PyTorch implementation. Does that help?

Comment (catli) [reply model/gru.py:77]: @zhampel  Thanks Zig! I tried the multi-class cross-entropy and it performed worse than MSE for some reason.
